Source/WebCore/style/StyleInvalidationFunctions.h

```diff
@@ -90,8 +90,11 @@ inline void traverseRuleFeatures(Element& element, TraverseFunction&& function)
     traverseRuleFeaturesForSlotted(element, function);
 
     // Ensure that the containing tree resolver also exists so it doesn't get created in the middle of invalidation.
-    if (element.isInShadowTree() && element.containingShadowRoot())
-        Style::Scope::forNode(*element.containingShadowRoot()->host()).resolver();
+    if (element.isInShadowTree() && element.containingShadowRoot()) {
+        auto& host = *element.containingShadowRoot()->host();
+        if (host.isConnected())
+            Style::Scope::forNode(host).resolver();
```

Comment:
Crash log that I'm fixing:

```
ASSERTION FAILED: node.isConnected()
style/StyleScope.cpp(189) : static WebCore::Style::Scope &WebCore::Style::Scope::forNode(WebCore::Node &)
1   0x1353ac510 WTFCrash
2   0x280601d14 WebCore::JSANGLEInstancedArrays::createPrototype(JSC::VM&, WebCore::JSDOMGlobalObject&)
3   0x2857051ac WebCore::Style::Scope::forNode(WebCore::Node&)
4   0x285657098 void WebCore::Style::traverseRuleFeatures<WebCore::Style::AttributeChangeInvalidation::invalidateStyle(WebCore::QualifiedName const&, WTF::AtomString const&, WTF::AtomString const&)::$_0>(WebCore::Element&, WebCore::Style::AttributeChangeInvalidation::invalidateStyle(WebCore::QualifiedName const&, WTF::AtomString const&, WTF::AtomString const&)::$_0&&)
5   0x285656dbc WebCore::Style::AttributeChangeInvalidation::invalidateStyle(WebCore::QualifiedName const&, WTF::AtomString const&, WTF::AtomString const&)
6   0x28395c690 WebCore::Style::AttributeChangeInvalidation::AttributeChangeInvalidation(WebCore::Element&, WebCore::QualifiedName const&, WTF::AtomString const&, WTF::AtomString const&)
7   0x2839427fc WebCore::Style::AttributeChangeInvalidation::AttributeChangeInvalidation(WebCore::Element&, WebCore::QualifiedName const&, WTF::AtomString const&, WTF::AtomString const&)
8   0x283939d38 WebCore::Element::setAttributeInternal(unsigned int, WebCore::QualifiedName const&, WTF::AtomString const&, WebCore::Element::SynchronizationOfLazyAttribute)
9   0x28393a298 WebCore::Element::setAttributeWithoutSynchronization(WebCore::QualifiedName const&, WTF::AtomString const&)
10  0x28406d830 WebCore::TextControlInnerTextElement::updateInnerTextElementEditabilityImpl(bool, bool)
11  0x283e83400 WebCore::TextControlInnerTextElement::updateInnerTextElementEditability(bool)
12  0x283e83348 WebCore::HTMLTextFormControlElement::updateInnerTextElementEditability()
13  0x283e83284 WebCore::HTMLTextFormControlElement::disabledStateChanged()
14  0x283da7528 WebCore::HTMLInputElement::disabledStateChanged()
15  0x283f0bb94 WebCore::ValidatedFormListedElement::setAncestorDisabled(bool)
16  0x283f0cd08 WebCore::ValidatedFormListedElement::removedFromAncestor(WebCore::Node::RemovalType, WebCore::ContainerNode&)
17  0x283d76418 WebCore::HTMLFormControlElement::removedFromAncestor(WebCore::Node::RemovalType, WebCore::ContainerNode&)
18  0x283dab310 WebCore::HTMLInputElement::removedFromAncestor(WebCore::Node::RemovalType, WebCore::ContainerNode&)
19  0x2837f03f8 WebCore::notifyNodeRemovedFromDocument(WebCore::ContainerNode&, WebCore::TreeScopeChange, WebCore::Node&)
20  0x2837f0234 WebCore::notifyChildNodeRemoved(WebCore::ContainerNode&, WebCore::Node&)
21  0x2837ec6f4 WebCore::ContainerNode::removeNodeWithScriptAssertion(WebCore::Node&, WebCore::ContainerNode::ChildChange::Source)
22  0x2837e9860 WebCore::ContainerNode::removeChild(WebCore::Node&)
23  0x2837e8e6c WebCore::ContainerNode::removeSelfOrChildNodesForInsertion(WebCore::Node&, WTF::Vector<WTF::Ref<WebCore::Node, WTF::RawPtrTraits<WebCore::Node> >, 11ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&)
24  0x2837ea9a8 WebCore::ContainerNode::appendChildWithoutPreInsertionValidityCheck(WebCore::Node&)
25  0x2837ed468 WebCore::ContainerNode::appendChild(WebCore::Node&)
26  0x283a1ae28 WebCore::Node::appendChild(WebCore::Node&)
27  0x28145a5a0 WebCore::jsNodePrototypeFunction_appendChildBody(JSC::JSGlobalObject*, JSC::CallFrame*, WebCore::JSNode*)::'lambda'()::operator()() const
28  0x28145a3e8 void WebCore::invokeFunctorPropagatingExceptionIfNecessary<WebCore::jsNodePrototypeFunction_appendChildBody(JSC::JSGlobalObject*, JSC::CallFrame*, WebCore::JSNode*)::'lambda'()>(JSC::JSGlobalObject&, JSC::ThrowScope&, WebCore::jsNodePrototypeFunction_appendChildBody(JSC::JSGlobalObject*, JSC::CallFrame*, WebCore::JSNode*)::'lambda'()&&)
29  0x28145a318 WebCore::jsNodePrototypeFunction_appendChildBody(JSC::JSGlobalObject*, JSC::CallFrame*, WebCore::JSNode*)
30  0x28145a0cc long long WebCore::IDLOperation<WebCore::JSNode>::call<&(WebCore::jsNodePrototypeFunction_appendChildBody(JSC::JSGlobalObject*, JSC::CallFrame*, WebCore::JSNode*)), (WebCore::CastedThisErrorBehavior)0>(JSC::JSGlobalObject&, JSC::CallFrame&, char const*)
31  0x281455054 WebCore::jsNodePrototypeFunction_appendChild(JSC::JSGlobalObject*, JSC::CallFrame*)
com.apple.WebKit.WebContent.Development terminated (pid 81791) for reason: crash
LEAK: 1 WebPageProxy
```

Before this PR, the node's connected flag wasn't yet set (erroneously I would say):

```cpp
void HTMLFormControlElement::removedFromAncestor(RemovalType removalType, ContainerNode& oldParentOfRemovedTree)
{
    bool wasMatchingInvalidPseudoClass = willValidate() && !isValidFormControlElement();

    m_validationMessage = nullptr;
    if (m_disabledByAncestorFieldset)
        setAncestorDisabled(computeIsDisabledByFieldsetAncestor()); // Shadow DOM modified

    bool wasInsideDataList = false;
    if (m_dataListAncestorState == InsideDataList) {
        m_dataListAncestorState = Unknown;
        wasInsideDataList = true;
    }

    HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree); // IsConnected flag removed
    FormListedElement::elementRemovedFromAncestor(*this, removalType);

    if (wasMatchingInvalidPseudoClass)
        removeInvalidElementToAncestorFromInsertionPoint(*this, &oldParentOfRemovedTree);

    if (wasInsideDataList)
        updateWillValidateAndValidity();
}
```

I'm not an expert on style invalidation, but IMO it's fine to check connectedness since we only ensuring resolver.

Reply:
Yeah, that makes sense to me.